src/xtgeo/xyz/_xyz_io.py

```diff
@@ -11,140 +14,372 @@
+    while XYZ differs somewhat.
 
-    self._df = pd.read_csv(
-        pfile,
-        delim_whitespace=True,
-        skiprows=0,
-        header=None,
-        names=[self._xname, self._yname, zname],
-        dtype=np.float64,
-        na_values=999.00,
-    )
+    The are some subtle differences between a RMS text file and XYZ file
+    according to RMS documentation:
+
+    Points:
+
+    rms_ascii and XYZ seems identical!
+
+    Polygons:
+
+    rms_ascii have 3 columns and will contain a line 999.0 999.0 999.0 to
+    seperate polygons.
+
+    The XYZ files valid for Polygons will have an ID column. It may be 3 or 4 columns,
+    if only 3 then the Z values is assumed to be 0.0. The last column is always the ID.
+    """
+    assume_rms_text = True  # assume that XYZ file is irap text based
+
+    ncolumns = 3
+    with open(pfile.file, "r", encoding="utf8") as stream:
+        fields = stream.readline().split()
+        if len(fields) == 4:
+            assume_rms_text = False
+            ncolumns = 4
+            if not is_polygons:
+                raise TypeError(
+                    "Points is assumed but this appears to be a XYZP "
+                    "for Polygons as the file has 4 columns"
+                )
+        else:
+            if not is_polygons:
+                assume_rms_text = True
+            else:
+                assume_rms_text = False
+
+            if "." in fields[2]:
+                assume_rms_text = True
+
+    kwargs = {}
+    _xn = kwargs["xname"] = "X_UTME"
+    _yn = kwargs["yname"] = "Y_UTMN"
+    _zn = kwargs["zname"] = zname
+
+    if assume_rms_text:
+        dfr = pd.read_csv(
+            pfile.file,
+            delim_whitespace=True,
+            skiprows=0,
+            header=None,
+            names=[_xn, _yn, _zn],
+            dtype=np.float64,
+            na_values=999.00,
+        )
+    if not is_polygons:
+        dfr.dropna(inplace=True)
+
+    # pylint: disable=unsupported-assignment-operation
+    if is_polygons and assume_rms_text:
+        # make a new polygon for every input line which are undefined (999 in input)
+        _pn = kwargs["pname"] = "POLY_ID"
+        dfr[_pn] = dfr.isnull().all(axis=1).cumsum().dropna()
+        dfr.dropna(axis=0, inplace=True)
+        dfr.reset_index(inplace=True, drop=True)
+
+    if is_polygons and not assume_rms_text:
+        _pn = kwargs["pname"] = "POLY_ID"
+        dtypes = {_xn: np.float64, _yn: np.float64, _pn: np.int32}
+        if ncolumns == 4:
+            dtypes = {_xn: np.float64, _yn: np.float64, _zn: np.float64, _pn: np.int32}
+
+        dfr = pd.read_csv(
+            pfile.file,
+            delim_whitespace=True,
+            skiprows=0,
+            header=None,
+            names=dtypes.keys(),
+            dtype=dtypes,
+            na_values=999.00,
+        )
+        if ncolumns == 3:
+            dfr.insert(2, _pn, 0.0)
+
+    kwargs["values"] = _ValidDataFrame(dfr)
```

Comment:
dfr will be unassigned if `not is_polygons and not assume_rms_text`.

Reply:
`assume_rms_text` will always be true for points